go-controller/pkg/node/gateway_shared_intf.go

```diff
@@ -214,42 +292,54 @@ func (npw *nodePortWatcher) createLbAndExternalSvcFlows(service *kapi.Service, s
 	// to the nodeIP / nodeIP:port of the host networked backend.
 	// And then ensure that return traffic is UnDNATed correctly back
 	// to the ingress / external IP
-	if epHostLocal {
+	isServiceTypeETPLocal := util.ServiceExternalTrafficPolicyLocal(service)
+	if isServiceTypeETPLocal && hasLocalHostNetworkEp {
+		// case1 (see function description for details)
 		var nodeportFlows []string
 		klog.V(5).Infof("Adding flows on breth0 for %s Service %s in Namespace: %s since ExternalTrafficPolicy=local", ipType, service.Name, service.Namespace)
-		// table 0, This rule matches on all traffic with dst ip == LoadbalancerIP / extenalIP, DNAT's it to the correct NodeIP
+		// table 0, This rule matches on all traffic with dst ip == LoadbalancerIP / externalIP, DNAT's it to the correct NodeIP
 		// If ipv6 make sure to choose the ipv6 node address for rule
 		if strings.Contains(flowProtocol, "6") {
 			nodeportFlows = append(nodeportFlows,
 				fmt.Sprintf("cookie=%s, priority=110, in_port=%s, %s, %s=%s, tp_dst=%d, actions=ct(commit,zone=%d,nat(dst=[%s]:%s),table=6)",
-					cookie, npw.ofportPhys, flowProtocol, nwDst, ipAddress, svcPort.Port, HostNodePortCTZone, npw.gatewayIPv6, svcPort.TargetPort.String()))
+					cookie, npw.ofportPhys, flowProtocol, nwDst, externalIPOrLBIngressIP, svcPort.Port, HostNodePortCTZone, npw.gatewayIPv6, svcPort.TargetPort.String()))
 		} else {
 			nodeportFlows = append(nodeportFlows,
 				fmt.Sprintf("cookie=%s, priority=110, in_port=%s, %s, %s=%s, tp_dst=%d, actions=ct(commit,zone=%d,nat(dst=%s:%s),table=6)",
-					cookie, npw.ofportPhys, flowProtocol, nwDst, ipAddress, svcPort.Port, HostNodePortCTZone, npw.gatewayIPv4, svcPort.TargetPort.String()))
+					cookie, npw.ofportPhys, flowProtocol, nwDst, externalIPOrLBIngressIP, svcPort.Port, HostNodePortCTZone, npw.gatewayIPv4, svcPort.TargetPort.String()))
 		}
 		nodeportFlows = append(nodeportFlows,
 			// table 6, Sends the packet to the host
 			fmt.Sprintf("cookie=%s, priority=110, table=6, actions=output:LOCAL",
-				cookie),
+				etpSvcOpenFlowCookie),
 			// table 0, Matches on return traffic, i.e traffic coming from the host networked pod's port, and unDNATs
 			fmt.Sprintf("cookie=%s, priority=110, in_port=LOCAL, %s, tp_src=%s, actions=ct(commit,zone=%d nat,table=7)",
 				cookie, flowProtocol, svcPort.TargetPort.String(), HostNodePortCTZone),
 			// table 7, the packet back out eth0 to the external client
 			fmt.Sprintf("cookie=%s, priority=110, table=7, "+
-				"actions=output:%s", cookie, npw.ofportPhys))
+				"actions=output:%s", etpSvcOpenFlowCookie, npw.ofportPhys))
 
 		npw.ofm.updateFlowCacheEntry(key, nodeportFlows)
 
+	} else if isServiceTypeETPLocal && !hasLocalHostNetworkEp && config.Gateway.Mode == config.GatewayModeLocal {
+		// case2 (see function description for details)
+		// will hit this only if 1)gatewaymode=local & 2)svc.externaltrafficpolicy=local & 3)we don't have
```

Comment:
i would suggest removing lines 326-328. I'm not sure where 3a-c are coming from and I dont think it adds more confusion than clarity.

Reply:
done.